p2p/src/peer_manager/peerdb_common/storage.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2021-2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub trait TransactionRo {
+    fn close(self);
+}
+
+pub trait TransactionRw {
+    fn abort(self);
+
+    fn commit(self) -> Result<(), storage::Error>;
+}
+
+pub trait Transactional<'t> {
+    /// Associated read-only transaction type.
+    type TransactionRo: TransactionRo + 't;
+
+    /// Associated read-write transaction type.
+    type TransactionRw: TransactionRw + 't;
+
+    /// Start a read-only transaction.
+    fn transaction_ro<'s: 't>(&'s self) -> Result<Self::TransactionRo, storage::Error>;
+
+    /// Start a read-write transaction.
+    fn transaction_rw<'s: 't>(&'s self) -> Result<Self::TransactionRw, storage::Error>;
+}
+
+/// Declare a trait that extends [`Transactional`] and adds additional constraints to its
+/// associated types `TransactionRo` and `TransactionRw`.
+/// ```
+/// # use p2p::{decl_storage_trait, peer_manager::peerdb_common::storage::Transactional};
+/// # trait StorageRead {}
+/// # trait StorageWrite {}
+/// decl_storage_trait!(Storage, StorageRead, StorageWrite);
+/// ```
+/// is basically equivalent to
+/// ```
+/// # use p2p::peer_manager::peerdb_common::storage::Transactional;
+/// # trait StorageRead {}
+/// # trait StorageWrite {}
+/// pub trait Storage: for<'t> Transactional<'t> + Send
+/// where
+///     for<'t> <Self as Transactional<'t>>::TransactionRo: StorageRead,
+///     for<'t> <Self as Transactional<'t>>::TransactionRw: StorageWrite,
+/// {
+/// }
+/// ```
+/// except that with the `where` clause you'd have to repeat the trait bounds for
+/// `TransactionRo`/`TransactionRw` everywhere where you use `Storage`, and with
+/// `decl_storage_trait` you don't.
+#[macro_export]
+macro_rules! decl_storage_trait {
```

Comment:
We're in the process of getting rid of the macros in chainstate storage. Please, avoid using macros. Debugging with macros is hell on earth.